### DISCUSSION TITLE
CPU voltage

### DISCUSSION THREAD

**Discussion by Bharath-VCG** [oshi/oshi#1860] [Ideas]
Hi, I'm wondering why there is no feature for retrieving processor input voltage?

I did some google. I found some solutions.

For Linux machines, there is a Linux command-line tool(s-tui) that is written in python that provides the CPU input voltage on the console.
For windows, there is an option to get it through jSensor lib. That feature was not even implemented on jSensor as well. I did see the debug log of jSensor. They are reading all information through OpenHardWareMonitor DLL file. So I saw that it is providing CPU voltage. But they haven't implemented it (I think they haven't found a solution for Linux).

Note: I was in need of CPU voltage. So, I implemented all the above on my project. For Linux, I rewrote s-tui logic on Java. For Linux, as jSensor handles most of the rest part. I just added a few bean classes and initialized them then finally appended them to the result. It seems very simple. I thought this should be added to OSHI. So I shared my thoughts.

Lib/tools I talked about:
https://github.com/profesorfalken/jSensors
https://github.com/amanusk/s-tui

Thanks,
Bharath

**Comment by dbwiddis** [comment]
OSHI does have a `Sensors` class which is supposed to provide such information: https://oshi.github.io/oshi/oshi-core/apidocs/oshi/hardware/Sensors.html

It does its best to attempt to read built-in information, but the reality is that most operating systems don't provide this information by default.  

Windows/WMI has some information but it's not populated the vast majority of the time, particularly in more recent hardware. The main problem is that reading the values is very dependent on the firmware/bios provider.  Some of the libraries you might see that read this info internally have dozens of special-cases.  See https://github.com/openhardwaremonitor/openhardwaremonitor/tree/master/Hardware, there are multiple different CPUs, or other heat monitoring applications supported.  It's simply not maintainable to attempt to copy every single one of those into OSHI!  See https://github.com/oshi/oshi/issues/119

As you mention, OpenHardwareMonitor does provide this information which is what jSensor uses.  OpenHardwareMonitor does publish its information to a WMI table that OSHI attempts to read, so in theory, if you are running OHM in the background on Windows, the `getVolts()` method should give you a useful result.

For Linux, we do attempt to read data from `/sys/class/hwmon` and/or `/sys/class/thermal`.  See https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxSensors.java
